erigon-lib/state/state_changeset.go

```diff
@@ -0,0 +1,113 @@
+package state
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"sort"
+
+	lru "github.com/hashicorp/golang-lru/v2"
+	"github.com/ledgerwatch/erigon-lib/common"
+	"github.com/ledgerwatch/erigon-lib/kv"
+)
+
+const MaxFastChangesets = 64
+
+type StateDiffKind uint8
+
+type StateChangeSet struct {
+	BeginTxIndex uint64                        // Txn index to unwind to
+	Diffs        [kv.DomainLen]StateDiffDomain // there are 4 domains of state changes
+}
+
+type KVPair struct {
+	Key   []byte
+	Value []byte
+}
+
+// StateDiffDomain represents a domain of state changes.
+type StateDiffDomain struct {
+	// We can probably flatten these into single slices for GC/cache optimization
+	keys          map[string][]byte
+	prevValues    map[string][]byte
+	keysSlice     []KVPair
+	prevValsSlice []KVPair
+}
+
+// RecordDelta records a state change.
+func (d *StateDiffDomain) DomainUpdate(key1, key2, prevValue, stepBytes []byte, prevStep uint64) {
+	if d.keys == nil {
+		d.keys = make(map[string][]byte)
+	}
+	if d.prevValues == nil {
+		d.prevValues = make(map[string][]byte)
+	}
+	prevStepBytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(prevStepBytes, ^prevStep)
+
+	key := append(key1, key2...)
```

Comment:
modify external `key1` is dangerous. maybe something like:
if len(key2) > 0 {
key1 = append(comon.Copy(key1), key2...)
}
key := string(comon.Copy(key1), stepBytes...))